Implements workaround for java.util.Date bug described in issue 160

dlurton opened the pull request:
See https://github.com/amzn/ion-java/issues/160.  Also there are ample comments in the diff.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

pbcornell commented on test/software/amazon/ion/TimestampTest.java:
```diff
@@ -860,6 +860,57 @@ public void testNewTimestampFromLong()
+                assertEquals(1, timestampPlusOne.getYear());
+
+                // Don't attempt to test this last case since it would create a timestamp that is before year 1.
+                if(tz.getRawOffset() >= 0) break;
```
I think you mean `continue` instead of `break`.  My preference would be:

```java
if(tz.getRawOffset() < 0) {
    Timestamp timestampMinusOne = ...;
    assertEquals(...);
}
```

dlurton replied:
Indeed.  Will be addressed.

pbcornell commented on test/software/amazon/ion/TimestampTest.java:
```diff
@@ -860,6 +860,57 @@ public void testNewTimestampFromLong()
+     * Regression test for https://github.com/amzn/ion-java/issues/160
+     */
+    @Test
+    public void testNewTimestampFromLongYearOneRegressionBug()
```
"FromLongYearOneRegressionBug" ?  perhaps just "FromYearOneRegressionBug"?

dlurton replied:
Yes.  Not sure where "Long" came from.

pbcornell commented on src/software/amazon/ion/Timestamp.java:
```diff
@@ -276,14 +277,33 @@ private void apply_offset(int offset)
 
-        // These fields are in the system timezone!
-        this._year    = checkAndCastYear(date.getYear() + 1900);
+        // java.util.Date has a Date(long) constructor that expects an epoch time in milliseconds. Date's getYear(),
```
I think we can simplify this comment;  what do you think about the following?

```
// https://github.com/amzn/ion-java/issues/160
// The java.util.Date(long) constructor expects an epoch time in milliseconds, and getYear(), getMonth(),
// getHour() on the resulting Date are supposed to return values adjusted to the default timezone.
// In Pacific Standard Time (offset -08:00), for a Date constructed with an epoch time equivalent to
// 0001-01-01T00:00:00.000Z, the Date.get*() methods should return values for
// 0000-12-31T16:00:00.000Z;  however, Date.getYear() incorrectly returns 1 in this scenario.
// The following if/else block compensates for this bug:
```

dlurton replied:
It actually returns -1899 (the year -1900) for some reason unfathomable to me.

I've modified this slightly:
```
// https://github.com/amzn/ion-java/issues/160
// The java.util.Date(long) constructor expects an epoch time in milliseconds, and getYear(), getMonth(),
// getHour() on the resulting Date are supposed to return values adjusted to the default timezone.
// In Pacific Standard Time (offset -08:00), for a Date constructed with an epoch time equivalent to
// 0001-01-01T00:00:00.000Z, the Date.get*() methods should return values for
// 0000-12-31T16:00:00.000Z;  however, Date.getYear() incorrectly returns a value for year 1 (-1899)
// in this scenario. The following if/else block compensates for this bug:
```

dlurton commented:
I've added [additional JDKs](https://github.com/amzn/ion-java/commit/c4750dfbb6072e467919365636af8ca31d0f9f59) to `.travis.yml`, rebased and pushed again and [Travis still reports no issues](https://travis-ci.org/amzn/ion-java/builds/440805728).
